File: src/julia/modules/Enumerations/src/StopRule.jl
"""
	@enum StopRule

Controls stop criteria. Stops either when:
- a `TARGET` value is found;
- or a `MAXTIME` is reached;
- or a `ITERATIONS` limit is reached;
- or the `FEASIBILITY` is achieved.
"""
@enum StopRule begin
	TARGET = 1
	MAXTIME = 2
	ITERATIONS = 3
	FEASIBILITY = 4
end

"""
	function parse(::Type{StopRule}, value::String)::StopRule

Parse `value` into a `StopRule`.
"""
function parse(::Type{StopRule}, value::String)::StopRule
	local_value = uppercase(strip(value)[1])
	if local_value == 'T'
		return TARGET
	elseif local_value == 'M'
		return MAXTIME
	elseif local_value == 'I'
		return ITERATIONS
	elseif local_value == 'F'
		return FEASIBILITY
	end

	throw(ArgumentError("cannot parse $value as StopRule"))
end

File: src/python/modules/enumerations/stop_rule.py
from enum import Enum


class StopRule(Enum):
    TARGET = 1
    MAXTIME = 2
    ITERATIONS = 3
    FEASIBILITY = 4

    @staticmethod
    def parse(value: str) -> 'StopRule':
        """
        Parse a string into a StopRule.
        T -> TARGET
        M -> MAXTIME
        I -> ITERATIONS
        F -> FEASIBILITY
        """
        local_value = value.strip().upper()[0]
        if local_value == 'T':
            return StopRule.TARGET
        elif local_value == 'M':
            return StopRule.MAXTIME
        elif local_value == 'I':
            return StopRule.ITERATIONS
        elif local_value == 'F':
            return StopRule.FEASIBILITY
        else:
            raise ValueError(f"Cannot parse '{value}' as StopRule.")
